armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+        self.keep_results = keep_results
+
+    def get_arg_names(self):
+        return list(self.arg_index)
+
+    def clear(self):
+        """
+        Clear all of the current values
+        """
+        self.values = [None] * len(self.arg_index)
+        self.values_set = [False] * len(self.arg_index)
+        self.batches = [None] * len(self.arg_index)
+
+    def add_writer(self, writer):
+        """
+        Add a writer to emit outputs to.
+        """
+        if writer not in self.writers:
+            self.writers.append(writer)
+
+    def set(self, name, value, batch):
+        """
+        Set the value for the arg name
+        """
+        if name not in self.arg_index:
+            raise ValueError(f"{name} not a valid arg name from {self.arg_index}")
+        i = self.arg_index[name]
+        self.values[i] = value
+        self.values_set[i] = True
+        self.batches[i] = batch
+        if self.auto_measure and self.is_ready():
+            self.measure()
+
+    def is_ready(self, raise_error=False):
+        """
+        Return True if all values have been set and batch numbers match
+            if raise_error is True, raise ValueError instead of returning False
+        """
+        if not all(self.values_set):
+            if raise_error:
+                raise ValueError(f"Not all values have been set: {self.values_set}")
+            return False
+        if any(self.batches[0] != batch for batch in self.batches):
+            if raise_error:
+                raise ValueError("Batch numbers are mismatched: {self.batches}")
+            return False
+        return True
+
+    def measure(self, clear_values=True):
+        self.is_ready(raise_error=True)
+        result = self.metric(*self.values, **self.metric_kwargs)
+        record = (self.name, self.batches[0], result)
+        if self.keep_results:
+            # Assume metric is sample-wise, but computed on a batch of samples
+            self._results.extend(record)
+            # TODO: only global?
+        for writer in self.writers:
+            writer.write(record)
+        if not self.keep_results and not self.writers and not self._warned:
+            log.warning(
+                f"Meter {self.name} has no writer added and keep_results is False"
+            )
+            self._warned = True
+        if clear_values:
+            self.clear()
+
+    def finalize(self):
+        """
+        Primarily intended for metrics of metrics, like mean or stdev of results
+        """
+        if self.final is None:
+            return
+
+        result = self.final(self.results, **self.final_kwargs)
+        record = (self.final_name, None, result)
+        self.final_result = result
+        for writer in self.writers:
+            writer.write(record)
+
+    def results(self):
+        if not self.keep_results:
+            raise ValueError("keep_results is False")
+        return self._results
```

Comment:
should `self._results` be list of 3-tuples instead of a flat list? Not sure what useful things we could do with the following format I'm seeing:

```
Pdb) self._results
	['sum of x_adv', 0, 52.496500501843975, 'sum of x_adv', 0, 57.73304135870708]
```

Reply:
Not sure what is most helpful. I meant this to be a list of the third value (the "result"), not a list of the records.
```
[51.74264097777119, 56.87368698209825, 61.65885794852083, ...
```

Reply:
Fixed.

Reply:
Got it, wasn't sure if there was a reason you were storing the strings